pubsub/awspubsub/awspubsub.go

```diff
@@ -0,0 +1,182 @@
+// Subscription: *sqs.SQS
+package awspubsub
+
+import (
+	"context"
+	"encoding/json"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/sns"
+	"github.com/aws/aws-sdk-go/service/sqs"
+	"gocloud.dev/pubsub"
+	"gocloud.dev/pubsub/driver"
+)
+
+type topic struct {
+	client *sns.SNS
+	arn    string
+}
+
+// TopicOptions will contain configuration for topics.
+type TopicOptions struct{}
+
+// OpenTopic opens the topic on AWS SNS for the given SNS client and topic ARN.
+func OpenTopic(ctx context.Context, client *sns.SNS, topicARN string, opts *TopicOptions) *pubsub.Topic {
+	return pubsub.NewTopic(openTopic(ctx, client, topicARN))
+}
+
+// openTopic returns the driver for OpenTopic. This function exists so the test
+// harness can get the driver interface implementation if it needs to.
+func openTopic(ctx context.Context, client *sns.SNS, topicARN string) driver.Topic {
+	return &topic{client: client, arn: topicARN}
+}
+
+// SendBatch implements driver.Topic.SendBatch.
+func (t *topic) SendBatch(ctx context.Context, dms []*driver.Message) error {
+	for _, dm := range dms {
+		attrs := map[string]*sns.MessageAttributeValue{}
+		for k, v := range dm.Metadata {
+			attrs[k] = &sns.MessageAttributeValue{
+				DataType:    aws.String("String"),
+				StringValue: aws.String(v),
+			}
+		}
+		req, _ := t.client.PublishRequest(&sns.PublishInput{
+			Message:           aws.String(string(dm.Body)),
+			MessageAttributes: attrs,
+			TopicArn:          &t.arn,
+		})
+		if err := req.Send(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// IsRetryable implements driver.Topic.IsRetryable.
+func (t *topic) IsRetryable(error) bool {
+	// The client handles retries.
+	return false
+}
+
+// As implements driver.Topic.As.
+func (t *topic) As(i interface{}) bool {
+	c, ok := i.(**sns.SNS)
+	if !ok {
+		return false
+	}
+	*c = t.client
+	return true
+}
+
+type subscription struct {
+	client *sqs.SQS
+	qURL   string
+}
+
+// SubscriptionOptions will contain configuration for subscriptions.
+type SubscriptionOptions struct{}
+
+// OpenSubscription opens a on AWS SQS for the given SQS client and queue URL.
+// The queue is assumed to be subscribed to some SNS topic, though there is no
+// check for this.
+func OpenSubscription(ctx context.Context, client *sqs.SQS, qURL string, opts *SubscriptionOptions) *pubsub.Subscription {
+	return pubsub.NewSubscription(openSubscription(ctx, client, qURL))
+}
+
+// openSubscription returns a driver.Subscription.
+func openSubscription(ctx context.Context, client *sqs.SQS, qURL string) driver.Subscription {
+	return &subscription{client: client, qURL: qURL}
+}
+
+// ReceiveBatch implements driver.Subscription.ReceiveBatch.
+func (s *subscription) ReceiveBatch(ctx context.Context, maxMessages int) ([]*driver.Message, error) {
+	output, err := s.receiveMessages()
+	if err != nil {
+		return nil, err
+	}
+	var ms []*driver.Message
+	for _, m := range output.Messages {
+		type Attribute struct {
```

Comment:
Since you never use this type, you can omit it and define the struct inline on line 124.

Reply:
Done.